packages/nextjs/src/server/clerkClient.ts

```diff
@@ -27,6 +33,54 @@ const clerkClient = createClerkClient({
     disabled: TELEMETRY_DISABLED,
     debug: TELEMETRY_DEBUG,
   },
-});
+};
+
+const createClerkClientWithOptions: typeof createClerkClient = options =>
+  createClerkClient({ ...clerkClientDefaultOptions, ...options });
+
+/**
+ * @deprecated
+ * This singleton is deprecated and will be removed in a future release. Please use `clerkClient()` as a function instead.
+ */
+const clerkClientSingleton = createClerkClient(clerkClientDefaultOptions);
+
+/**
+ * Constructs a BAPI client that accesses request data within the runtime.
+ * Necessary if middleware dynamic keys are used.
+ */
+const clerkClientForRequest = () => {
+  let requestData;
+
+  try {
+    const request = buildRequestLike();
+    const encryptedRequestData = getHeader(request, constants.Headers.ClerkRequestData);
+    requestData = decryptClerkRequestData(encryptedRequestData);
+  } catch (err) {
+    // When outside of middleware runtime, fallbacks to access request data from `NextRequest`
```

Comment:
I _think_ this comment is outdated based on the logic 👀  We're falling back to using our internal async-local store.

Reply:
@LauraBeatris I'm wondering if we need to check the error here to make sure it's not something we should re-throw. I think it's possible that `headers()` and `cookies()` throw errors related to static generation bail-out that might need to be rethrown...

Reply:
@BRKalow I've restructured to check if the store is defined first, instead of relying on the error being thrown to determine the runtime. 

```ts
const middlewareStore = clerkMiddlewareRequestDataStore.getStore();
if (Object.values(middlewareStore ?? {}).length) {
  requestData = middlewareStore;
} else {
  // When outside of middleware runtime, fallbacks to access request data from `NextRequest`
  const request = buildRequestLike();
  const encryptedRequestData = getHeader(request, constants.Headers.ClerkRequestData);
  requestData = decryptClerkRequestData(encryptedRequestData);
}
```